File: fileExample/folder3/LogicsImpl.scala
package u04lab.polyglot.a01b
import scala.jdk.javaapi.OptionConverters
import u04lab.polyglot.{OptionToOptional, Pair}
import u04lab.code.Option
import u04lab.code.Option.*
import u04lab.code.List
import u04lab.code.List.*
import scala.util.Random

/** solution and descriptions at https://bitbucket.org/mviroli/oop2019-esami/src/master/a01b/sol2/ */

private trait Cell:
  def position: Pair[Int, Int]
  def hasMine: Boolean
  def hasMine_=(state: Boolean): Unit
  def isAdjacent(cell: Cell): Boolean

object Cell:
  def apply(position: Pair[Int, Int]): Cell =
    CellImpl(position, false)

  //Cosa cambia se metto override o no?
  //Cosa cambia tra val, var, def?
  //Cosa comporta riscrivere la def con _ indicando state?
  case class CellImpl(position: Pair[Int, Int], override var hasMine: Boolean) extends Cell :
    override def isAdjacent(cell: Cell): Boolean =
      Math.abs(position.getX - cell.position.getX) <= 1 &&
        Math.abs(position.getY - cell.position.getY) <= 1


class LogicsImpl(private val size: Int, private val mines: Int) extends Logics:

  private var cells: List[Cell] = empty
  //private var random: Random = Random(size * size)
  private var hitCells: Int = 0

  //private def getRandomPosition(): Pair[Int, Int] = Pair(random.nextInt(size), random.nextInt(size))
  private def numberOfAdjacentMine(cell: Cell): Int =
    length(filter(cells)(c => c.isAdjacent(cell) && c.hasMine))

  for
    i <- 0 until size
    j <- 0 until size
  do
    cells = append(cells, cons(Cell(Pair(i, j)), empty))

  List.apply(List.getRandoms(cells)(mines)){_.hasMine = true}
  def hit(x: Int, y: Int): java.util.Optional[Integer] =
    val cell: Option[Cell] = List.get(filter(cells)(c => c.position.equals(Pair(x, y))), 0)
    cell match
      case None() => java.util.Optional.empty()
      case Some(c) if c.hasMine => java.util.Optional.empty()
      case Some(c) if !c.hasMine => {
        hitCells = hitCells + 1
        java.util.Optional.of(numberOfAdjacentMine(c))
      }

  def won: Boolean = hitCells + mines == length(cells)

File: fileExample/folder3/LogicsImpl.java
package e1;

import java.util.*;

public class LogicsImpl implements Logics {
	
	private final Pair<Integer,Integer> pawn;
	private Pair<Integer,Integer> knight;
	private final Random random = new Random();
	private final int size;
	private MoveStrategy moveStrategy = new MoveStrategyImpl();
	 
    public LogicsImpl(int size){
    	this.size = size;
        this.pawn = this.randomEmptyPosition();
        this.knight = this.randomEmptyPosition();	
    }
	public LogicsImpl(int size, int knightRow, int knightCol){
		this.size = size;
		if (knightRow<0 || knightCol<0 || knightRow >= this.size || knightCol >= this.size) {
			throw new IndexOutOfBoundsException();
		}
		this.pawn = this.randomEmptyPosition();
		this.knight = new Pair<>(knightCol, knightRow);
	}
    
	private final Pair<Integer,Integer> randomEmptyPosition(){
    	Pair<Integer,Integer> pos = new Pair<>(this.random.nextInt(size),this.random.nextInt(size));
    	// the recursive call below prevents clash with an existing pawn
    	return this.pawn!=null && this.pawn.equals(pos) ? randomEmptyPosition() : pos;
    }
    
	@Override
	public boolean hit(int row, int col) {
		/*if (row<0 || col<0 || row >= this.size || col >= this.size) {
			throw new IndexOutOfBoundsException();
		}

		return this.moveStrategy.moveKnight(row,col,size);
		// Below a compact way to express allowed moves for the knight
		int x = row-this.knight.getX();
		int y = col-this.knight.getY();
		if (x!=0 && y!=0 && Math.abs(x)+Math.abs(y)==3) {
			this.knight = new Pair<>(row,col);
			return this.pawn.equals(this.knight);
		}*/
		return false;
	}

	@Override
	public boolean hasKnight(int row, int col) {
		return this.knight.equals(new Pair<>(row,col));
	}

	@Override
	public boolean hasPawn(int row, int col) {
		return this.pawn.equals(new Pair<>(row,col));
	}
}
